src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -228,265 +216,100 @@ where
 /// about. This second stream will be used by the `remap_operator` to mint new
 /// timestamp bindings into the remap shard.
-fn source_reader_operator<G, C, R>(
-    scope: &G,
+fn source_reader_operator<G, C>(
+    scope: &mut G,
     config: RawSourceCreationConfig,
     source_connection: C,
     connection_context: ConnectionContext,
-    resume_uppers: impl futures::Stream<Item = Antichain<R::Time>> + 'static,
+    resume_uppers: impl futures::Stream<Item = Antichain<C::Time>> + 'static,
 ) -> (
-    Collection<G, Result<SourceMessage<R::Key, R::Value>, SourceReaderError>, R::Diff>,
+    Collection<G, Result<SourceMessage<C::Key, C::Value>, SourceReaderError>, Diff>,
     Stream<G, Infallible>,
     Stream<G, (WorkerId, HealthStatusUpdate)>,
     Rc<dyn Any>,
 )
 where
-    G: Scope<Timestamp = R::Time>,
-    C: SourceConnectionBuilder<Reader = R> + 'static,
-    R: SourceReader,
+    G: Scope<Timestamp = C::Time>,
+    C: SourceRender + 'static,
 {
-    let RawSourceCreationConfig {
-        name,
-        id,
-        num_outputs: _,
-        worker_id,
-        worker_count,
-        timestamp_interval,
-        encoding,
-        storage_metadata: _,
-        resume_upper: _,
-        source_resume_upper,
-        base_metrics,
-        now: _,
-        persist_clients: _,
-        source_statistics,
-        shared_remap_upper: _,
-    } = config;
+    let source_id = config.id;
+    let worker_id = config.worker_id;
+    let source_statistics = config.source_statistics.clone();
 
-    let mut builder = AsyncOperatorBuilder::new(name.clone(), scope.clone());
+    let resume_uppers = resume_uppers.inspect(move |upper| {
+        let upper = upper.pretty();
+        trace!(%upper, "timely-{worker_id} source({source_id}) received resume upper");
+    });
 
-    let info = builder.operator_info();
-    // TODO(guswynn): should sources still be able to self-activate?
-    // probably not, so we should remove this
-    let sync_activator = scope.sync_activator_for(&info.address[..]);
+    let (data, progress, health, token) =
+        source_connection.render(scope, config, connection_context, resume_uppers);
 
-    let (mut output, stream) = builder.new_output();
-    // We don't need the output because we simply downgrade the output's capability.
-    let (_upper_output, upper_stream) = builder.new_output();
-    let (mut health_output, health_stream) = builder.new_output();
+    let mut builder = AsyncOperatorBuilder::new("Inspector".to_owned(), scope.clone());
 
-    let button = builder.build(move |mut capabilities| async move {
-        let health_cap = capabilities.pop().unwrap();
-        let upper_cap = capabilities.pop().unwrap();
-        let data_cap = capabilities.pop().unwrap();
+    let mut data_input = builder.new_input(&data.inner, Pipeline);
+    let (mut data_output, data) = builder.new_output();
+    let (mut _progress_output, derived_progress) = builder.new_output();
```

Comment:
so we drop its output capability so its frontier exactly follows the input frontier?